### PR TITLE
Added logic to track affected properties on models

### PR DESCRIPTION
**Pull request by vlady777**
Added logic to have possibility to track model's properties that were affected during request

### CODE REVIEW

File: Util/Model/AffectedPropertiesTrackableTrait.php [hunk 1]
```diff
@@ -0,0 +1,46 @@
...
+ * file that was distributed with this source code.
+ */
+namespace Mango\Bundle\JsonApiBundle\Util\Model;
+
+/**
+ * Trait to work with affected properties
+ *
+ * @author Vlad Yarus <vladislav.yarus@intexsys.lv>
+ */
+trait AffectedPropertiesTrackableTrait
```
Comment (dbalabka) [inline Util/Model/AffectedPropertiesTrackableTrait.php:15]: @vlady777 you are able to cover this functionality with functional tests. Please see \Mango\Bundle\JsonApiBundle\Tests\Serializer\DeserializationTest

File: Serializer/JsonApiDeserializationVisitor.php [hunk 2]
```diff
@@ -60,11 +60,14 @@ public function visitProperty(PropertyMetadata $metadata, $data, Context $contex
...
+            $visit = false;
             $relationship = [];
-            if (isset($data['data']['relationships'][$propertyName]['data'])) {
+            if (array_key_exists('data', $data['data']['relationships'][$propertyName])) {
                 $relationship = $data['data']['relationships'][$propertyName]['data'];
-            } elseif (isset($data['relationships'][$propertyName]['data'])) {
+                $visit = true;
+            } elseif (array_key_exists('data', $data['relationships'][$propertyName])) {
                 $relationship = $data['relationships'][$propertyName]['data'];
+                $visit = true;
```
Comment (dbalabka) [inline Serializer/JsonApiDeserializationVisitor.php:70]: @vlady777  you are able to cover this functionality with functional tests. Please see \Mango\Bundle\JsonApiBundle\Tests\Serializer\DeserializationTest

File: Serializer/Accessor/DefaultAccessorStrategy.php [hunk 3]
```diff
@@ -0,0 +1,33 @@
...
+     */
+    public function setValue($object, $value, PropertyMetadata $metadata)
+    {
+        parent::setValue($object, $value, $metadata);
+
+        if ($object instanceof AffectedPropertiesAwareInterface) {
+            $object->addAffectedProperty($metadata->name);
+        }
+    }
+}
```
Comment (dbalabka) [inline Serializer/Accessor/DefaultAccessorStrategy.php]: new line

File: Util/Model/AffectedPropertiesAwareInterface.php [hunk 4]
```diff
@@ -0,0 +1,31 @@
...
+     */
+    public function addAffectedProperty(string $propertyName);
+
+    /**
+     * Returns all affected property names
+     *
+     * @return array
+     */
+    public function getAffectedProperties(): array;
+}
```
Comment (dbalabka) [inline Util/Model/AffectedPropertiesAwareInterface.php]: new line

File: Util/Model/AffectedPropertiesTrackableTrait.php [hunk 5]
```diff
@@ -0,0 +1,46 @@
...
+    /**
+     * Returns all affected property names
+     *
+     * @return array
+     */
+    public function getAffectedProperties(): array
+    {
+        return $this->affectedProperties;
+    }
+}
```
Comment (dbalabka) [inline Util/Model/AffectedPropertiesTrackableTrait.php]: new line

File: Serializer/Accessor/DefaultAccessorStrategy.php [hunk 6]
```diff
@@ -0,0 +1,33 @@
...
+use JMS\Serializer\Metadata\PropertyMetadata;
+use Mango\Bundle\JsonApiBundle\Util\Model\AffectedPropertiesAwareInterface;
+
+/**
+ * DefaultAccessorStrategy is an override for original DefaultAccessorStrategy to add
+ * custom behavior like track affected properties, etc.
+ *
+ * @author Vlad Yarus <vladislav.yarus@intexsys.lv>
+ */
+class DefaultAccessorStrategy extends BaseAccessor
```
Comment (dbalabka) [inline Serializer/Accessor/DefaultAccessorStrategy.php:20]: @vlady777  you are able to cover this functionality with functional tests. Please see \Mango\Bundle\JsonApiBundle\Tests\Serializer\DeserializationTest

**Review by dbalabka — CHANGES_REQUESTED** [review]

File: Util/Model/AffectedPropertiesTrackableTrait.php (hunk 1 above)
Comment (vlady777) [reply Util/Model/AffectedPropertiesTrackableTrait.php:15]: Ticket was created for that. See related tickets.

File: Serializer/JsonApiDeserializationVisitor.php (hunk 2 above)
Comment (vlady777) [reply Serializer/JsonApiDeserializationVisitor.php:70]: Ticket was created for that. See related tickets.

File: Serializer/Accessor/DefaultAccessorStrategy.php (hunk 3 above)
Comment (vlady777) [reply Serializer/Accessor/DefaultAccessorStrategy.php]: Added

File: Util/Model/AffectedPropertiesAwareInterface.php (hunk 4 above)
Comment (vlady777) [reply Util/Model/AffectedPropertiesAwareInterface.php]: Added

File: Util/Model/AffectedPropertiesTrackableTrait.php (hunk 5 above)
Comment (vlady777) [reply Util/Model/AffectedPropertiesTrackableTrait.php]: Added

File: Serializer/Accessor/DefaultAccessorStrategy.php (hunk 6 above)
Comment (vlady777) [reply Serializer/Accessor/DefaultAccessorStrategy.php:20]: Ticket was created for that. See related tickets.

**Review by dbalabka — APPROVED** [review]

**Pull request merged**